pubsub/gossipsub/gossipsub-v2.0.md

```diff
@@ -0,0 +1,171 @@
+
+- If the router decides to forward the message eagerly, it will just forward the full message to that mesh peer.
+- If the router decides to forward the message lazily, it will send `IANNOUNCE` with the message id attached instead to tell them that it just receives a new message. If they want the full content of the message, they should send `INEED` back.
+
+After processing the message payload, the router will process the new control messages as follows:
+
+- Upon receiving `IANNOUNCE`, the router checks the [`seen` cache][seen-cache] to see if it has been seen or not. If not, it adds the sending peer to `acache[msgid]`, where `msgid` is the one attached to `IANNOUNCE`. If `acache[msgid]` was previously empty, the router pops `acache[msgid]` and sends `INEED` to that peer immediately.
+
+  After the router sends `INEED`, it will time out if it doesn't receive the message back in time, as indicated by `timeout`. If the timeout happens, the router will pop `acache[msgid]` send `INEED` to the next peer. If it still times out, keep going with next peers until the cache runs out of peers.
+
+  Notice that timeouts can delay receiving the message. It's worth noting that timeouts are rare, since the peer just sent the router `IANNOUNCE` so it means the network condition is already good and the peer should be able to send the actual message without problems.
+
+  In gossipsub v2.1, we plan to penalize peers that don't send the message in time.
+
+- Upon receiving `INEED` after sending `IANNOUNCE`, the router must send the full content back without an option to decline. Otherwise it will be deemed misbehaving.
+
+Apart from forwarding received messages, the router can of course publish messages on its own behalf. This is very similar to forwarding received messages. It also has to toss a coin to decide whether to send the message eagerly or lazily.
+
+[seen-cache]: ../gossipsub/gossipsub-v1.0.md#message-cache
+
+### Message Publishing
+
+For message publishing, as long as `D_announce` is less than `D`, full messages are published to our mesh peers. There is no advantage to lazy mesh
+propagation as none of the peers have seen the message before. In the event `D_announce` is equivalent to `D` we disable publishing the full message as 
+the message originator is trivially identifiable if message propagation is completely announcement based. 
+
+### Protobuf
+
+The protobuf messages are identical to those specified in the [gossipsub v1.2.0 specification](https://github.com/libp2p/specs/blob/master/pubsub/gossipsub/gossipsub-v1.2.md) with the following control message modifications:
+
+```protobuf
+message RPC {
+    // ... see definition in the gossipsub specification
+}
+
+message ControlMessage {
+    // messages from v1.2
+    repeated ControlIAnnounce iannounce = 6;
+    repeated ControlINeed ineed = 7;
```

Comment:
can't we just do this with IWANT?
In fact we could just send IHAVEs instead of using INEED.

Reply:
The reasoning was that this would function very differently from `IHAVE/IWANT` which is emitted periodically. While message announcements were meant to always be immediate and be primarily be used for message propagation.

Reply:
uhm ok, fair enough.

We also need to consider the interplay of IDONTWANT and IANNOUNCE as well.

Reply:
We can just say that if we received IDONTWANT before, we won't send IANNOUNCE.